pkg/s3select/input.go

```diff
@@ -0,0 +1,356 @@
+		}
+	}
+
+	return row
+}
+
+// convertMySQL Replaces double quote escape for column names with backtick for
+// the MySQL parser
+func convertMySQL(random string) string {
+	return strings.Replace(random, "\"", "`", len(random))
+}
+
+// readHeader reads the header into the header variable if the header is present
+// as the first row of the csv
+func (reader *Input) readHeader() error {
+	var readErr error
+	if reader.options.HasHeader {
+		reader.firstRow, readErr = reader.reader.Read()
+		if readErr != nil {
+			return ErrCSVParsingError
+		}
+		reader.header = reader.firstRow
+		reader.firstRow = nil
+		reader.minOutputLength = len(reader.header)
+	} else {
+		reader.firstRow, readErr = reader.reader.Read()
+		reader.header = make([]string, len(reader.firstRow))
+		for i := 0; i < reader.minOutputLength; i++ {
+			reader.header[i] = strconv.Itoa(i)
+		}
+
+	}
+	return nil
+}
+
+// createStatXML is the function which does the marshaling from the stat
+// structs into XML so that the progress and stat message can be sent
+func (reader *Input) createStatXML() (string, error) {
+	if reader.options.Compressed != "" {
+		reader.stats.BytesProcessed += processSize(reader.header)
+	}
+	if !(reader.options.Compressed != "") {
+		reader.stats.BytesProcessed = reader.options.StreamSize
+		reader.stats.BytesScanned = reader.stats.BytesProcessed
+	}
+	statXML := &stats{
+		BytesScanned:   reader.stats.BytesScanned,
+		BytesProcessed: reader.stats.BytesProcessed,
+		BytesReturned:  reader.stats.BytesReturned,
+	}
+	out, err := xml.Marshal(statXML)
+	if err != nil {
+		return "", err
+	}
+	return xml.Header + string(out), nil
+}
+
+// createProgressXML is the function which does the marshaling from the progress structs into XML so that the progress and stat message can be sent
+func (reader *Input) createProgressXML() (string, error) {
+	if reader.options.HasHeader {
+		reader.stats.BytesProcessed += processSize(reader.header)
+	}
+	if !(reader.options.Compressed != "") {
+		reader.stats.BytesScanned = reader.stats.BytesProcessed
+	}
+	progressXML := &progress{
+		BytesScanned:   reader.stats.BytesScanned,
+		BytesProcessed: reader.stats.BytesProcessed,
+		BytesReturned:  reader.stats.BytesReturned,
+	}
+	out, err := xml.Marshal(progressXML)
+	if err != nil {
+		return "", err
+	}
+	return xml.Header + string(out), nil
+}
+
+// Header returns the header of the reader. Either the first row if a header
+// set in the options, or c#, where # is the column number, starting with 0.
+func (reader *Input) Header() []string {
+	return reader.header
+}
+
+// Row is a Struct for keeping track of key aspects of a row.
+type Row struct {
+	record string
+	err    error
+}
+
+// Execute is the function where all the blocking occurs, It writes to the HTTP
+// response writer in a streaming fashion so that the client can actively use
+// the results before the query is finally finished executing. The
+func (reader *Input) Execute(writer http.ResponseWriter) error {
```

Comment:
Let's change the function signature to a more generic writer

```
func (reader *Input) Execute(writer io.Writer) error {
```

I am expecting that we are not using any internal features of http.ResponseWriter